### ISSUE TITLE
用colab训练报错 'StreamToLoguru' object has no attribute 'isatty'

### ISSUE THREAD

**Issue by w-x-j00** [Megvii-BaseDetection/YOLOX#1437]

**Comment by bsense-rius** [comment]
Dear Sirs,

I am also facing the very same error! I have been running succesfully your excellent model in a colab notebook.

Since 2 days ago, without making any change to the code, when running train.py, when the trainer.py creates the tensorboard logger it complains about: 'StreamToLoguru' object has no attribute 'isatty'

```
!python tools/train.py \
    -f nano.py \
    -b 16 \
    --fp16 \
    -o \
    -c yolox_nano.pth
...
2022-07-15 06:29:36 | INFO     | pycocotools.coco:88 - index created!
2022-07-15 06:29:38 | ERROR    | yolox.core.launch:98 - An error has been caught in function 'launch', process 'MainProcess' (370), thread 'MainThread' (140111377680256):
Traceback (most recent call last):

  File "/usr/local/lib/python3.7/dist-packages/tensorboard/compat/__init__.py", line 42, in tf
    from tensorboard.compat import notf  # noqa: F401

ImportError: cannot import name 'notf' from 'tensorboard.compat' (/usr/local/lib/python3.7/dist-packages/tensorboard/compat/__init__.py)


During handling of the above exception, another exception occurred:


Traceback (most recent call last):

  File "tools/train.py", line 141, in <module>
    args=(exp, args),
          │    └ Namespace(batch_size=16, cache=False, ckpt='yolox_nano.pth', devices=None, dist_backend='nccl', dist_url=None, exp_file='nano...
          └ ╒═══════════════════╤════════════════════════════════════════════════════════════════════════════════════════════════════════...

> File "/content/YOLOX/yolox/core/launch.py", line 98, in launch
    main_func(*args)
    │          └ (╒═══════════════════╤═══════════════════════════════════════════════════════════════════════════════════════════════════════...
    └ <function main at 0x7f6e36b88950>

  File "tools/train.py", line 118, in main
    trainer.train()
    │       └ <function Trainer.train at 0x7f6db00f2c20>
    └ <yolox.core.trainer.Trainer object at 0x7f6d972cfa10>

  File "/content/YOLOX/yolox/core/trainer.py", line 74, in train
    self.before_train()
    │    └ <function Trainer.before_train at 0x7f6d972c9b90>
    └ <yolox.core.trainer.Trainer object at 0x7f6d972cfa10>

  File "/content/YOLOX/yolox/core/trainer.py", line 181, in before_train
    self.tblogger = SummaryWriter(os.path.join(self.file_name, "tensorboard"))
    │               │             │  │    │    │    └ './YOLOX_outputs/nano'
    │               │             │  │    │    └ <yolox.core.trainer.Trainer object at 0x7f6d972cfa10>
    │               │             │  │    └ <function join at 0x7f6e36c367a0>
    │               │             │  └ <module 'posixpath' from '/usr/lib/python3.7/posixpath.py'>
    │               │             └ <module 'os' from '/usr/lib/python3.7/os.py'>
    │               └ <class 'torch.utils.tensorboard.writer.SummaryWriter'>
    └ <yolox.core.trainer.Trainer object at 0x7f6d972cfa10>

  File "/usr/local/lib/python3.7/dist-packages/torch/utils/tensorboard/writer.py", line 246, in __init__
    self._get_file_writer()
    │    └ <function SummaryWriter._get_file_writer at 0x7f6d9b9d8d40>
    └ <torch.utils.tensorboard.writer.SummaryWriter object at 0x7f6d95b07c10>

  File "/usr/local/lib/python3.7/dist-packages/torch/utils/tensorboard/writer.py", line 277, in _get_file_writer
    self.log_dir, self.max_queue, self.flush_secs, self.filename_suffix
    │    │        │    │          │    │           │    └ ''
    │    │        │    │          │    │           └ <torch.utils.tensorboard.writer.SummaryWriter object at 0x7f6d95b07c10>
    │    │        │    │          │    └ 120
    │    │        │    │          └ <torch.utils.tensorboard.writer.SummaryWriter object at 0x7f6d95b07c10>
    │    │        │    └ 10
    │    │        └ <torch.utils.tensorboard.writer.SummaryWriter object at 0x7f6d95b07c10>
    │    └ './YOLOX_outputs/nano/tensorboard'
    └ <torch.utils.tensorboard.writer.SummaryWriter object at 0x7f6d95b07c10>

  File "/usr/local/lib/python3.7/dist-packages/torch/utils/tensorboard/writer.py", line 76, in __init__
    log_dir, max_queue, flush_secs, filename_suffix
    │        │          │           └ ''
    │        │          └ 120
    │        └ 10
    └ './YOLOX_outputs/nano/tensorboard'

  File "/usr/local/lib/python3.7/dist-packages/tensorboard/summary/writer/event_file_writer.py", line 72, in __init__
    tf.io.gfile.makedirs(logdir)
    │                    └ './YOLOX_outputs/nano/tensorboard'
    └ <module 'tensorboard.compat.tf' via LazyModule (not yet loaded)>

  File "/usr/local/lib/python3.7/dist-packages/tensorboard/lazy.py", line 65, in __getattr__
    return getattr(load_once(self), attr_name)
                   │         │      └ 'io'
                   │         └ <module 'tensorboard.compat.tf' via LazyModule (not yet loaded)>
                   └ <function lazy_load.<locals>.wrapper.<locals>.load_once at 0x7f6db00eff80>

  File "/usr/local/lib/python3.7/dist-packages/tensorboard/lazy.py", line 97, in wrapper
    cache[arg] = f(arg)
    │     │      │ └ <module 'tensorboard.compat.tf' via LazyModule (not yet loaded)>
    │     │      └ <function lazy_load.<locals>.wrapper.<locals>.load_once at 0x7f6db00efef0>
    │     └ <module 'tensorboard.compat.tf' via LazyModule (not yet loaded)>
    └ {}

  File "/usr/local/lib/python3.7/dist-packages/tensorboard/lazy.py", line 50, in load_once
    module = load_fn()
             └ <function tf at 0x7f6db00efb90>

  File "/usr/local/lib/python3.7/dist-packages/tensorboard/compat/__init__.py", line 45, in tf
    import tensorflow

  File "/usr/local/lib/python3.7/dist-packages/tensorflow/__init__.py", line 51, in <module>
    from ._api.v2 import compat

  File "/usr/local/lib/python3.7/dist-packages/tensorflow/_api/v2/compat/__init__.py", line 37, in <module>
    from . import v1

  File "/usr/local/lib/python3.7/dist-packages/tensorflow/_api/v2/compat/v1/__init__.py", line 30, in <module>
    from . import compat

  File "/usr/local/lib/python3.7/dist-packages/tensorflow/_api/v2/compat/v1/compat/__init__.py", line 37, in <module>
    from . import v1

  File "/usr/local/lib/python3.7/dist-packages/tensorflow/_api/v2/compat/v1/compat/v1/__init__.py", line 47, in <module>
    from tensorflow._api.v2.compat.v1 import lite

  File "/usr/local/lib/python3.7/dist-packages/tensorflow/_api/v2/compat/v1/lite/__init__.py", line 9, in <module>
    from . import experimental

  File "/usr/local/lib/python3.7/dist-packages/tensorflow/_api/v2/compat/v1/lite/experimental/__init__.py", line 8, in <module>
    from . import authoring

  File "/usr/local/lib/python3.7/dist-packages/tensorflow/_api/v2/compat/v1/lite/experimental/authoring/__init__.py", line 8, in <module>
    from tensorflow.lite.python.authoring.authoring import compatible

  File "/usr/local/lib/python3.7/dist-packages/tensorflow/lite/python/authoring/authoring.py", line 43, in <module>
    from tensorflow.lite.python import convert

  File "/usr/local/lib/python3.7/dist-packages/tensorflow/lite/python/convert.py", line 29, in <module>
    from tensorflow.lite.python import util

  File "/usr/local/lib/python3.7/dist-packages/tensorflow/lite/python/util.py", line 51, in <module>
    from jax import xla_computation as _xla_computation

  File "/usr/local/lib/python3.7/dist-packages/jax/__init__.py", line 59, in <module>
    from .core import eval_context as ensure_compile_time_eval

  File "/usr/local/lib/python3.7/dist-packages/jax/core.py", line 47, in <module>
    import jax._src.pretty_printer as pp

  File "/usr/local/lib/python3.7/dist-packages/jax/_src/pretty_printer.py", line 56, in <module>
    CAN_USE_COLOR = _can_use_color()
                    └ <function _can_use_color at 0x7f69bb320320>

  File "/usr/local/lib/python3.7/dist-packages/jax/_src/pretty_printer.py", line 54, in _can_use_color
    return sys.stdout.isatty()
           │   └ <yolox.utils.logger.StreamToLoguru object at 0x7f6d9778aa90>
           └ <module 'sys' (built-in)>

AttributeError: 'StreamToLoguru' object has no attribute 'isatty'

**Comment by DaliaMahdy** [comment]
I have the same problem!

**Comment by bsense-rius** [comment]
I have been investigating about it.

The issue is a happens in the package jax/_src/pretty_printer.py, in the function can_use_color()
```
def _can_use_color() -> bool:
  try:
    # Check if we're in IPython or Colab
    ipython = get_ipython()  # type: ignore[name-defined]
    shell = ipython.__class__.__name__
    if shell == "ZMQInteractiveShell":
      # Jupyter Notebook
      return True
    elif "colab" in str(ipython.__class__):
      # Google Colab (external or internal)
      return True
  except NameError:
    pass
  # Otherwise check if we're in a terminal
  return sys.stdout.isatty()

CAN_USE_COLOR = _can_use_color(). 
```
THE PROBLEM IS THAT THE INSTRUCTION get_ipython() FOR WHATEVER UNKNOW REASON FOR ME IS NOT RECOGNIZED WITHIN! Then an exception is raised that forces to run return sys.stdout.isatty(). Due to the fact that Megv's Yolox utils/logger.py replaces stdout with StreamToLogUru, as this class does not have the function isatty(), the program suddenly stops.
 
Maybe a change of ipython version in colab that messed it up?

**Comment by bsense-rius** [comment]
It can be overriden by using -wandb logger, but it is very annoying that default behaviour is not working in colab.
:(

**Comment by jakevdp** [comment]
I think the issue is tha `StreamToLoguru`, defined in this package, is being used to patch `sys.stdout`, but does not define all the necessary methods of `sys.stdout`, in particular the `isatty` method. https://github.com/Megvii-BaseDetection/YOLOX/blob/48fd2a95837a92e0583492210b5ebeb02a4e7d7d/yolox/utils/logger.py#L34

**Comment by w-x-j00** [comment]
I still get an error, but I stumbled upon the !pip uninstall jax command and it works, I don't know what effect it will have

**Comment by bsense-rius** [comment]
Might be a good option. But it would be best if Megvii Team just add :

def isatty(): return True 

in class StreamLoguru at logger.py. That easy

**Comment by FateScript** [comment]
Thanks for your feed back, I will try to solve this issue in the week. Plz wait or work around this issue.

**Comment by w-x-j00** [comment]
Thanks to the method proposed by bsense-rius, I tried it, and it works, still looking forward to the solution from Megvii Team.

**Comment by FateScript** [comment]
Hi, we fix this issue in #1443 . We checked codelink posted by @jakevdp and decided to use solution provided by @bsense-rius .
We also note that a repo written by @Kazuhito00 is using the same solution.
Feel free to ask if you have any further quesiotn.
